graph/generate.go

```diff
@@ -0,0 +1,254 @@
+/*
+Package graph provides core lineage graph services. It defines the graph datatype and provides the "New"
+function for creating one.
+*/
+package graph
+
+import (
+	"fmt"
+
+	mbp "github.com/GoogleCloudPlatform/healthcare-data-harmonization/mapping_engine/proto"
+	proto "github.com/golang/protobuf/proto"
+)
+
+var idFactory IDfactory = &autoIncFactory{currentID: 0}
+
+// setIDfactory sets the interface used for ID generation
+func setIDfactory(factory IDfactory) {
+	idFactory = factory
+}
+
+// msgContext is a wrapper around a proto message storing additional context, like parent node.
+type msgContext struct {
+	msg        proto.Message
+	descendant Node
+}
+
+type stack []msgContext
```

Comment:
Does this need to be thread-safe?

Reply:
No parallelization optimizations are planned